source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,523 @@
+    /// Command without data transfer
+    Control,
+    /// Command for reading data, contains the number of bytes to read
+    Read(usize),
+    /// Command for writing data, contains the number of bytes to write
+    Write(usize),
+}
+
+/// The different types of responses that can be sent by the card
+#[derive(Debug, Eq, PartialEq)]
+pub enum ResponseType {
+    /// No Response
+    None,
+    /// Response with 48-bit length
+    Short,
+    /// Response with 48-bit length + check *busy* after response
+    ShortWithBusySignal,
+    /// Response with 136-bit length
+    Long,
+}
+
+/// Response returned by the card, can be short or long, depending on command
+#[must_use]
+pub enum Response {
+    /// The 32-bit value from the 48-bit response.
+    /// Potentially also includes the data buffer if this was a read command
+    Short {
+        /// The response on the command line
+        value: u32,
+        /// The received data, in case of a read command
+        data: Option<FixedVec<u8>>,
+    },
+    /// The 128-bit value from the 136-bit response
+    Long(u128),
+}
+
+/// Errors returned by the [`SdmmcService`]
+#[derive(Debug, Eq, PartialEq)]
+pub struct Error {
+    kind: ErrorKind,
+    message: Option<&'static str>,
+}
+
+/// The different types of errors that can occur
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// The service is currently busy and cannot handle the request
+    Busy,
+    /// Invalid or unexpected response was received
+    Response,
+    /// Invalid or unexpected data was received
+    Data,
+    /// The provided buffer does not meet the requirements
+    Buffer,
+    /// A timeout occurred
+    Timeout,
+    /// A different error has occurred
+    Other,
+}
+
+impl Default for Command {
+    fn default() -> Self {
+        Command {
+            index: 0,
+            argument: 0,
+            options: HardwareOptions::None,
+            kind: CommandKind::Control,
+            rsp_type: ResponseType::None,
+            rsp_crc: false,
+            buffer: None,
+        }
+    }
+}
+
+impl Error {
+    /// Create an error from a type and message
+    pub fn new(kind: ErrorKind, message: &'static str) -> Self {
+        Self {
+            kind,
+            message: Some(message),
+        }
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(value: ErrorKind) -> Self {
+        Self {
+            kind: value,
+            message: None,
+        }
+    }
+}
+
+impl core::fmt::Display for Error {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "SD/MMC error: ")?;
+        match self.kind {
+            ErrorKind::Busy => write!(f, "busy "),
+            ErrorKind::Response => write!(f, "response "),
+            ErrorKind::Data => write!(f, "data "),
+            ErrorKind::Buffer => write!(f, "buffer "),
+            ErrorKind::Timeout => write!(f, "timeout "),
+            ErrorKind::Other => write!(f, "other "),
```

Comment:
tiny style nit, not a blocker: i try to avoid using the `write!` macro when no formatted arguments are interpolated into the string. it's up to you whether you actually care about this, though... 
```suggestion
            ErrorKind::Data => f.write_str("data "),
            ErrorKind::Buffer => f.write_str("buffer "),
            ErrorKind::Timeout => f.write_str("timeout "),
            ErrorKind::Other => f.write_str("other "),
```